src/main/java/reactor/core/publisher/FluxReplay.java

```diff
@@ -45,6 +47,912 @@
+		public Throwable getError() {
+			return error;
+		}
+
+		@Override
+		public int capacity() {
+			return Integer.MAX_VALUE;
+		}
+
+		@Override
+		public void add(T value) {
+			int i = tailIndex;
+			Object[] a = tail;
+			if (i == a.length - 1) {
+				Object[] b = new Object[a.length];
+				b[0] = value;
+				tailIndex = 1;
+				a[i] = b;
+				tail = b;
+			}
+			else {
+				a[i] = value;
+				tailIndex = i + 1;
+			}
+			size++;
+		}
+
+		@Override
+		public void onError(Throwable ex) {
+			error = ex;
+			done = true;
+		}
+
+		@Override
+		public void onComplete() {
+			done = true;
+		}
+
+		void replayNormal(ReplaySubscription<T> rs) {
+			int missed = 1;
+
+			final Subscriber<? super T> a = rs.actual();
+			final int n = batchSize;
+
+			for (; ; ) {
+
+				long r = rs.requested();
+				long e = 0L;
+
+				Object[] node = (Object[]) rs.node();
+				if (node == null) {
+					node = head;
+				}
+				int tailIndex = rs.tailIndex();
+				int index = rs.index();
+
+				while (e != r) {
+					if (rs.isCancelled()) {
+						rs.node(null);
+						return;
+					}
+
+					boolean d = done;
+					boolean empty = index == size;
+
+					if (d && empty) {
+						rs.node(null);
+						Throwable ex = error;
+						if (ex != null) {
+							a.onError(ex);
+						}
+						else {
+							a.onComplete();
+						}
+						return;
+					}
+
+					if (empty) {
+						break;
+					}
+
+					if (tailIndex == n) {
+						node = (Object[]) node[tailIndex];
+						tailIndex = 0;
+					}
+
+					@SuppressWarnings("unchecked") T v = (T) node[tailIndex];
+
+					a.onNext(v);
+
+					e++;
+					tailIndex++;
+					index++;
+				}
+
+				if (e == r) {
+					if (rs.isCancelled()) {
+						rs.node(null);
+						return;
+					}
+
+					boolean d = done;
+					boolean empty = index == size;
+
+					if (d && empty) {
+						rs.node(null);
+						Throwable ex = error;
+						if (ex != null) {
+							a.onError(ex);
+						}
+						else {
+							a.onComplete();
+						}
+						return;
+					}
+				}
+
+				if (e != 0L) {
+					if (r != Long.MAX_VALUE) {
+						rs.produced(e);
+					}
+				}
+
+				rs.index(index);
+				rs.tailIndex(tailIndex);
+				rs.node(node);
+
+				missed = rs.leave(missed);
+				if (missed == 0) {
+					break;
+				}
+			}
+		}
+
+		void replayFused(ReplaySubscription<T> rs) {
+			int missed = 1;
+
+			final Subscriber<? super T> a = rs.actual();
+
+			for (; ; ) {
+
+				if (rs.isCancelled()) {
+					rs.node(null);
+					return;
+				}
+
+				boolean d = done;
+
+				a.onNext(null);
+
+				if (d) {
+					Throwable ex = error;
+					if (ex != null) {
+						a.onError(ex);
+					}
+					else {
+						a.onComplete();
+					}
+					return;
+				}
+
+				missed = rs.leave(missed);
+				if (missed == 0) {
+					break;
+				}
+			}
+		}
+
+		@Override
+		public void replay(ReplaySubscription<T> rs) {
+			if (!rs.enter()) {
+				return;
+			}
+
+			if (rs.fusionMode() == NONE) {
+				replayNormal(rs);
+			}
+			else {
+				replayFused(rs);
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 623:632, 858:867.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.